src/bitcoin/electrum/mod.rs

```diff
@@ -198,25 +200,42 @@ impl Electrum {
         if let Some(keychain_update) = keychain_update {
             self.bdk_wallet.apply_keychain_update(keychain_update);
         }
-        let changeset = self.bdk_wallet.apply_connected_chain_update(chain_update);
-
-        let mut changes_iter = changeset.into_iter();
-        let reorg_common_ancestor = loop {
-            match changes_iter.next() {
-                Some((height, Some(_))) => {
-                    // `BlockHash` being `Some(_)` means a checkpoint at this height was added to the chain.
-                    // Since we iterate in ascending height order, we'll see the lowest block height first.
-                    // If the lowest height it adds is higher than our height before syncing, we're good.
-                    // Else if it's adding a block at height before syncing or lower, it's a reorg.
-                    break if height > local_chain_tip.height() {
-                        None
-                    } else {
-                        log::info!("Block chain reorganization detected.");
-                        Some(self.bdk_wallet.find_block_at_or_before_height(height))
-                    };
+
+        let reorg_common_ancestor = if !is_rescanning {
+            let changeset = self.bdk_wallet.apply_connected_chain_update(chain_update);
+            let mut changes_iter = changeset.into_iter();
+            loop {
+                match changes_iter.next() {
+                    Some((height, Some(_))) => {
+                        // `BlockHash` being `Some(_)` means a checkpoint at this height was added to the chain.
+                        // Since we iterate in ascending height order, we'll see the lowest block height first.
+                        // If the lowest height it adds is higher than our height before syncing, we're good.
+                        // Else if it's adding a block at height before syncing or lower, it's a reorg.
+                        break if height > local_chain_tip.height() {
+                            None
+                        } else {
+                            log::info!("Block chain reorganization detected.");
+                            Some(self.bdk_wallet.find_block_at_or_before_height(height))
+                        };
+                    }
+                    Some((_, None)) => continue,
+                    None => break None,
                 }
-                Some((_, None)) => continue,
-                None => break None,
+            }
+        } else {
+            // In case of a full scan, just replace our local chain with the new one.
+            self.bdk_wallet.replace_chain_from_tip(chain_update);
```

Comment:
Why is this necessary? `apply_connected_chain_update` should have exactly the same effect if the only block in common is the genesis.

Reply:
That may be true, but my thinking is as follows: When we do the full scan, it's from the genesis block, and so the chain update we receive is with respect to the genesis block only and has no knowledge of our local chain tip. If there's a reorg, I'm worried that the chain update with respect to the genesis block may not correctly update our local chain.

Reply:
You mean if there was a reorg, then we rescan, and then arrive to this line, right? So a situation like:
- Our local chain is: `G` <- `A` <- `B` <- `C`
- The update is: `G` <- `D` <- `E` <- `F`

Then i see no reason why `apply_connected_chain_update` would not correctly update our local chain? It's exactly what it's designed to do.